### DISCUSSION TITLE
Are there ways to gain in *runtime* memory without losing too much on performances?

### DISCUSSION THREAD

**Discussion by masklinn** [rust-lang/regex#1206] [Q&A]
I've finally done some very basic benching [`regex-filtered`](https://github.com/ua-parser/uap-rust/tree/main/regex-filtered) against `FilteredRE2`: ua-parser/uap-rust@1af15c172aaec3eea6dd41a110b0e490a9c7f7ed

After [fixing a dumb mistake](https://github.com/ua-parser/uap-rust/commit/4f1c7df94f4171d78e0128146254d99e6a12c27e) it's basically on par in terms of runtime (and that's hot even its final form).

However there's an other standout issue: the RSS / memory footprint of the regex-filtered version is much higher than that of the re2 version, and simple tracking of the rss over phases using [memory-stats](https://crates.io/crates/memory-stats) shows it is largely attributable to `regex`, this was confirmed by just loading all the regexes into a `Vec<Regex>` and finding that the RSS climbs from ~2MB (after loading the arguments and reading the 633 regex strings from the file) to ~100MB (and then a bit further to ~140MB when matching, I assume that's because caches come into play then).

I looked at the different features of regex, and the only knob which seemed to influence RSS is `dfa`, which reduces RSS down to ~50MB before matching which is similar to re2 (though caches then tack ~ 20MB on when matching occurs). Sadly, it's also very much load bearing and disabling it makes runtime shoot up from 46 seconds to 157 seconds on my machine (for 100 iterations over 75158 strings, using 633 regexes).

And so I wanted to know, is there a knob I missed? Maybe something wrong I did with regex::bytes or something? Or is it the tradeoff? I assume `regex-lite` would also be an option but wouldn't really be competitive with re2?

In fairness I can't say it's of major importance, the runtime seems a lot more relevant, but it's useful information to stash away given the use case for `regex-filtered` would routinely be to load hundreds to thousands of regexes and it might inform e.g. lazy-compiling regexes (under the assumption that most of them will never actually be needed).

**Comment by BurntSushi** [comment]
> After [fixing a dumb mistake](https://github.com/ua-parser/uap-rust/commit/4f1c7df94f4171d78e0128146254d99e6a12c27e) it's basically on par in terms of runtime (and that's hot even its final form).

Out of curiosity, why not just [implement the sparse set yourself](https://github.com/rust-lang/regex/blob/8856fe36ac7dc37989e6ffb26b5fc57189bae626/regex-automata/src/util/sparse_set.rs)? It's not a lot of code and it's pretty simple.

As to your question... It's hard to say. If you could show a MRE Rust program that uses a lot more memory than an equivalent C++ program using RE2, then I could likely give you a better answer.

Otherwise, the main knob to turn is probably to disable Unicode mode. The regex crate and RE2 have different support for Unicode. While there's a lot of differences in the details, probably the biggest difference is that the regexes `\w`, `\d` and `\s` are all Unicode-aware in Rust but specifically ASCII-only in RE2. This means that things like `\w{100}` are _a lot_ bigger in Rust than in RE2.

Also, in a multi-threaded environment, I would generally expect the regex crate to use more memory than RE2. RE2's lazy DFA makes use one of one shared cache across all threads for a single lazy DFA, where as the regex crate uses one cache per thread. But if you aren't measuring memory in a multi-threaded benchmark, this difference shouldn't come up.

There could also be memory usage related bugs. For example, I haven't looked at https://github.com/rust-lang/regex/issues/1116 yet.

Otherwise it's very difficult to answer your question in the abstract. I would really want an MRE to give a more precise answer. (For example, maybe there's something different about how the NFAs are represented? Or maybe there's something causing a prefilter somewhere to be bigger than one might expect?) In general, there _is_ a lot of pressure on runtime performance in the form of benchmarks, but very little on memory usage. So there could very well be gains to be had here.

**Reply by masklinn** [reply]
> Out of curiosity, why not just [implement the sparse set yourself](https://github.com/rust-lang/regex/blob/8856fe36ac7dc37989e6ffb26b5fc57189bae626/regex-automata/src/util/sparse_set.rs)? It's not a lot of code and it's pretty simple.

I'm planning to do that but I wanted to make sure the issue was fixed first without the risk of adding more errors of my own. And then I noticed the difference in memory use remained after that fix so I figured I'd look at it a bit.

> As to your question... It's hard to say. If you could show a MRE Rust program that uses a lot more memory than an equivalent C++ program using RE2, then I could likely give you a better answer.

The bench scripts should be runnable easily, the sample files are included, for [the Rust version](https://github.com/ua-parser/uap-rust/blob/main/regex-filtered/examples/bench.rs) you should just need to 

    cargo r -r --example bench -- samples/devices.regexes samples/useragents.txt -r 100 -q

from that crate's root (I've got examples with the same name in both so can't be run from the repo root, I need to fix that).

The [C++ version](https://github.com/ua-parser/uap-rust/tree/main/regex-filtered/re2) is probably more involved (unless the flake works for you), and the CLI is a bit uglier for now:

    ./bench ../samples/devices.regexes ../samples/useragents.txt 100 1

But I can probably cut them down to just loading all the regexes in `Vec`/`std::vector` if you want, that is essentially the crux of the difference as far as I've been able to tell.

> Otherwise, the main knob to turn is probably to disable Unicode mode.

I tried that and didn't see a difference, although some of the regexes use `\d` so I had to keep `unicode-perl` enabled (I tried just bool or gencat enabled but I think I had regex parsing / compilation failures and didn't try much further).

> This means that things like `\w{100}` are a lot bigger in Rust than in RE2.

Yeah I saw that when building the prefilter. And now I'm remembering that while I tried `regex::bytes` I didn't disable `u` mode, so I checked its memory use with all the unicode stuff enabled didn't I? 🤦 

> But if you aren't measuring memory in a multi-threaded benchmark, this difference shouldn't come up.

Yeah nah the scripts are just basic single-threaded stuff so that shouldn't be a factor.

> There could also be memory usage related bugs. For example, I haven't looked at https://github.com/rust-lang/regex/issues/1116 yet.

Ah interesting. So I'll need to:

- test regex::bytes actually disabling unicode mode rather than leaving it enabled
- maybe take a better look at which bits of unicode I couldn't disable, possibly swap out the `\d` for `[0-9]`
- see what happens if I downgrade regex to 1.7, although the sample set makes *significant* use of capture groups, if the regression is only on non-capturing regexes it probably won't be a factor

>  In general, there is a lot of pressure on runtime performance in the form of benchmarks, but very little on memory usage.

Yeah completely understandable, to be honest I probably wouldn't have noticed if I hadn't used `time -l` as a rough yardstick. Not to mention in most programs with a handful of regexes it would get lost in the rest of the stuff and even a 100% overhead wouldn't be very visible, having hundreds of regexes and little else is definitely an outlier situation.

Thanks.

**Reply by BurntSushi** [reply]
Yeah I think checking Unicode mode is the best possible thing here. Of course, if things like `\w` and `\d` are used only sparingly, then disabling Unicode mode probably won't do much. Be on the lookout for things like `\w{100}` though.

And by disabling Unicode mode, yes, I do indeed mean `(?-u)` or `RegexBuilder::unicode(false)`. Whether you enable or disable the crate features is just going to impact binary size, but, by design, cannot impact whether a regex uses them or not. (Adding or removing crate features can only increase or decrease the set of valid regexes. They can never change the semantics of a valid regex.) Disabling Unicode mode should always be possible without error when using `regex::bytes`. Just be careful that things like `[^a]` will match single bytes like `\xFF` (always invalid UTF-8) or even `\xE2` (invalid UTF-8 on its own, but is a valid prefix).

For making an MRE, one possibility is to start trimming down the set of regexes to see how that impact memory usage. Maybe there is a Pareto distribution here or something.

**Reply by masklinn** [reply]
Do you use a linux or a mac box for development? (or both / either) I'm asking because getting the RSS is platform-specific, and while memory-stats does that for me in rust that's not the case in C++.

**Reply by BurntSushi** [reply]
Linux.

**Reply by BurntSushi** [reply]
(I do have an M2 mac mini with a dev environment setup, but I greatly prefer Linux. I only use mac and Windows when necessary.)

**Reply by BurntSushi** [reply]
I saw a reply here by email but don't see it here. I'm on mobile, but I would suggest wrapping the global allocator if you want more precise measurements. Otherwise you're probably just going to be doing battle with the optimizations performed by the allocator itself. See how dtolnay did it here: https://github.com/rust-lang/regex/issues/1059#issuecomment-1666584529

**Reply by masklinn** [reply]
I've hit a bit of a pickle I think, memory-stats turned out to be too imprecise (or a warning) to get good data so I procrasticircled a bit before ending up at jemalloc where I should have started, however I'm hitting issues with very odd numbers: in order to isolate measurements I first figured I'd check them one at a time (running a measurement program per) measuring memory before and after but the numbers seemed excessive, so I changed it up back to a loop measuring before/after creating the regex, and got a bunch of regexes taking no or negative sizes. For instance here's the start of the multiline script:

```
     1335200
     1680864   +345664  ((?:SCH)\-[A-Za-z0-9 ]{1,50})(/?[^ ]*|) Build
     1911040   +230176  (ATP[0-9]{4})(?: Build|\) AppleWebKit)
     2369792   +458752  (BoxNotes)/(\d+)\.(\d+)\.(\d+)
     2369792        +0  (Dar)(win)/(10).(\d+).*\((?:i386|x86_64)\)
     2726976   +357184  (Dar)(win)/(11).(\d+).{0,100}\((?:i386|x86_64)\)
     2262464   -464512  (Dar)(win)/(12).(\d+).{0,100}\((?:i386|x86_64)\)
     1629856   -632608  (Dar)(win)/(13).(\d+).{0,100}\((?:i386|x86_64)\)
     1644944    +15088  (Dar)(win)/(9).(\d+).{0,100}\((?:i386|x86_64|Power Macintosh)\)
```
Column 1 is [`stats.allocated`](https://jemalloc.net/jemalloc.3.html#stats.allocated), column 2 is change from previous iteration, all regexes get added to a preallocated vec after compilation so they're not dropped. Regexes 7, 8, and 9 (darwin 11 to 13) are identical yet memory goes down more after the second one than they go up before the first. I also have a hard time believing the 4th one would need no memory. It's more credible for some of the set which is literal, but even here while it's common it's not actually consistent:
```
    37813424        +0 (PlayStation Portable)
    37818224     +4800 (PlayStation Vita)
    37789424    -28800 (Play[Bb]ook)
    37789424        +0 (Player FM)$
    37881584    +92160 (Player FM|Pocket Casts|DoggCatcher|Spotify|MediaMonkey|MediaGo|BashPodder)
    37881584        +0 (Pocket Casts)$
[...]
    46203536    +45824 (WinNT4.0)
    46213936    +10400 (Windows (?:NT 5\.2|NT 5\.1))
    46213936        +0 (Windows 10)
    46213936        +0 (Windows ?CE)
    46216336     +2400 (Windows ?Mobile)
    46232720    +16384 (Windows NT 10\.0)
    46253200    +20480 (Windows NT 5\.0)
    46298256    +45056 (Windows NT 6\.2; ARM;)
    46302256     +4000 (Windows NT 6\.3; ARM;)
[...]
   178337192        +0 compatible; monitis
   178337192        +0 iPhone
   178337192        +0 i\-mate ([A-Za-z0-9]+)
   178337192        +0 linux
   178339496     +2304 rekonq
```

I figure it's because regex might use compilation / expression caches internally and those interfere with measurements? But I'm not sure if that's true, or how to interact with that, there's a `perf-cache` feature but per its documentation it seems to do nothing nowadays, and to deal more with *runtime* caching (which I'm not hitting as I never actually run the regexes). I'm not too clear on what regex-automata does either as it does mention a memory pool and `regex_automata::hybrid::regex::Cache`, but that looks to be runtime concern for the lazy automata.

In case that worked any better I tried measuring the memory reduction when *dropping* the regex object, but that doesn't seem to fare much better: out of 1183 regexes 480 release no memory on drop, and I get strange bucketings clustered around power-of-two multiples of 64KiB but there are a lot of odd values (a bunch of 48KiB, lots around 103 and 105, and a majority of 260 rather than an even 256):

```
       0 ============================================== /480/ ===============================================
   14400 =
   49152 ============== 31 =============
   55552 =
   65536 =========================================== 89 ==========================================
   98304 =====
  106304 =================================== 74 ===================================
  107904 ============================================== /151/ ===============================================
  131072 ============================================== /144/ ===============================================
  155456 ==
  157056 =
  180224 =
  255360 =======
  262144 =========
  266560 ============================================== /105/ ===============================================
  269760 =====
  271360 =
  311296 ==
  315712 ==
  318912 ==
  320512 =
  322112 =
  525120 ====================
  526720 ==
  531552 =
  531648 ====
  531744 ===
  536320 =
  544640 =
  544736 =
  549984 =
  552496 =
  552720 =
  556000 =
  573440 =
  593504 =
  630592 =
  632192 =
  677088 =
  682048 =
  684224 =
  763584 =
  790848 =
 1097728 ===
 1099328 =
 1106240 ===
 1107520 =
 1120640 =
 1134528 =
 1153344 =
 1154880 =
 1198400 ==
 1315136 ===
 2136256 =
 2320896 =
 2341376 ==
 8321920 =
```

I'm not really sure what I'm missing or how to progress from there.

**Reply by masklinn** [reply]
> I saw a reply here by email but don't see it here. 

Yeah sorry I posted an early version of the comment but figured I'd forgotten to take note of the old "accumulative" version of the bench script, so I went to rewrite it before posting it again.

> I'm on mobile, but I would suggest wrapping the global allocator if you want more precise measurements. Otherwise you're probably just going to be doing battle with the optimizations performed by the allocator itself. See how dtolnay did it here: [#1059 (comment)](https://github.com/rust-lang/regex/issues/1059#issuecomment-1666584529)

It's my understanding that that's what jemalloc stats are supposed to surface. But I can always try using a more simplistic proxy allocator, given dtolnay posted the entire thing. I'll give that a shot.

**Reply by masklinn** [reply]
> I saw a reply here by email but don't see it here. I'm on mobile, but I would suggest wrapping the global allocator if you want more precise measurements. Otherwise you're probably just going to be doing battle with the optimizations performed by the allocator itself.

Dammit from the documentation (and friends' suggestions) I assumed jemalloc's stats.allocated would be... the allocated stuff. But apparently it's doing weird things behind the scenes indeed. Custom allocators do seem to be a lot more reliable.

I put up https://github.com/masklinn/remem which provides for the regexes dataset and test scripts. I tried to get the evaluation / testing programs similar enough that comparisons mostly hold, but I'll readily admit that my C++ is really far from good. Also some of the behaviours I don't understand (e.g. in some cases `regex` needs no allocations for captures, in others it does, and I don't know what the difference is).

Anyway setting asides the unicode stuff, it looks like:

- regex has a much higher baseline, even comparing regex::bytes with `(?-u)` to re2 (utf8 and latin1)
-  bounded repetitions (e.g. `{1, 100}` rather than `+`) has a really big impact, especially combined with captures in unicode mode

e.g. two of the initial entries are:
|       re2 utf8      ||       re2 latin1     | |  regex unicode/string ||   regex ascii/string  ||   regex ascii/bytes   || |
|-|-|-|-|-|-|-|-|-|-|-|
| compile  |  capture  |  compile  |  capture  |  compile  |  capture  |  compile  |  capture  |  compile  |  capture  | expression|
|   2200   |    1876   |    2080   |    1620   |   16394   |     0     |     -     |     -     |    8466   |     0     |  (Dar)(win)/(10).(\d+).*\((?:i386|x86_64)\)|
|   9248   |   16532   |    5076   |    6772   |   141946  |     0     |     -     |     -     |   21626   |     0     |  (Dar)(win)/(11).(\d+).{0,100}\((?:i386|x86_64)\)|

So
- re2 in latin/ascii mode has a 2x multiplier from `*` to `{0, 100}`
- re2 in utf8 mode is 4x
- regex in ascii/bytes mode is 2.5x
- regex in unicode/string mode is 8.5x

With a simpler case (and ignoring the capture memory)
| |.\*|(.\*)|.{0,500}|(.{0,500})|
|-|-:|-:|-:|-:|
|re2 - utf8|808|896 (x1.1)|92636 (x115)|92692 (x115)|
|re2 - ascii|648|720 (x1.11)|18620 (x29)|18676 (x29)|
|regex - unicode|7784|11944 (x1.5)|655568 (x84)|1704216 (x219)|
|regex - binascii|6160|6544 (x1.05)|70896 (x11.5)|87352 (x16)|

On this one specifically regex actually does quite well on repetitions except for the repetition + capture unicode version, however since its baseline is 10x that of re2 it's still quite a ways behind.

**Reply by masklinn** [reply]
For what it's worth, although the upstream of what I'm using regex-filtered for does not specify their regex semantics the reference implementation is in javascript which uses strict ASCII for their perl-style character classes, and as you wrote up in the initial answer re2 also documents that they use strict ASCII semantics for the same. So while not touching regex-filtered I added a downgrade step to the sub-project, which converts `\d`, `\D`, `\w`, and `\W` to the ASCII sets documented by MDN / re2 (`\s`/`\S` doesn't seem significantly affected by the issue). Thanks much for regex supporting nested enumerated character classes incidentally, that made the conversion a lot less fraught as I need to care a lot less about context for this conversion. 

Through log-diving I also confirmed that the extensive use of bounded repetitions is not actually to bound repetitions but to mitigate catastrophic backtracking concerns, so while that's a lot iffier the downgrade filter also converts "large" bounded repetitions to unbounded (as it is my understanding that finite automata engines like regex are not affected).

This results in a ~50% reduction in peak memory footprint on processing a sample file, and despite the additional prefiltering (and allocations) a ~10% reduction in runtime.

Feel free to close this discussion if you don't see a point to keeping it open.

PS: no idea what could be done about the amplification caused by bounded repetitions and captures, but maybe `RegexBuilder` could grow a flag to make perl-style character class work off of ascii semantics? That would make equivalent conversions (and honest but misguided comparisons) from engines with lesser unicode support slightly easier not to screw up. I know that regex supports ascii-only classes via the POSIX-style classes, but Perl-style classes seem better supported (e.g. neither JS nor Python's re support POSIX-style). For prior art, Python has [`re.ASCII`](https://docs.python.org/3/library/re.html#re.ASCII) which gimps perl-style character classes down to ASCII semantics.

I know that `regex::bytes` can be used for some of it, because it only works on bytes in textual contexts it's a bit of a downgrade.

**Reply by BurntSushi** [reply]
With respect to Perl style classes being ASCII, that's already supported: just disable Unicode mode. Either globally, or locally within the regex: `(?-u:\w)` is equivalent to `[[:word:]]`. Perhaps what you want is a mode that _only_ downgrades the Perl style classes to ASCII and leaves everything else be. I'm unsure about that personally.

The simplifications you made to your regexes sound good to me. I had no idea that the bounds were added not for semantics, but to deal with backtrackers. That is interesting!

I do think a more careful analysis of the memory footprint differences between this crate and RE2 is warranted, but I don't currently consider it higher priority.

**Reply by masklinn** [reply]
> With respect to Perl style classes being ASCII, that's already supported: just disable Unicode mode. Either globally, or locally within the regex: (?-u:\w) is equivalent to [[:word:]]. Perhaps what you want is a mode that only downgrades the Perl style classes to ASCII and leaves everything else be.

Yes, that is indeed what I meant as the upstream in this cases uses `.` a fair bit (and in general I assume people would commonly find that in existing regexes for projects they're trying to move over to `regex`). I don't know if there are other metacharacters with that issue in the set, but I know that one has tripped me pretty much every time I've tried to use the `-u` with the string version of the API.

>  I had no idea that the bounds were added not for semantics, but to deal with backtrackers. That is interesting!

Yeah I had no idea either, I only found out when I tried to log dive to see if I could find explanations of the semantics requirements, and stumbled on a commit converting a bunch of unbounded repetitions to bounded ones.
